tokio/src/runtime/builder.rs

```diff
@@ -168,7 +172,36 @@ impl Builder {
     /// # }
     /// ```
     pub fn thread_name(&mut self, val: impl Into<String>) -> &mut Self {
-        self.thread_name = val.into();
+        let val = val.into();
+        self.thread_name = Arc::new(Box::new(move || val.clone()));
+        self
+    }
+
+    /// Set the name fn of threads spawned by the `Runtime`'s thread pool.
+    ///
+    /// The default name fn is `|| "tokio-runtime-worker".into()`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::runtime;
+    /// # use std::sync::atomic::{AtomicUsize, Ordering};
+    ///
+    /// # pub fn main() {
+    /// let rt = runtime::Builder::new()
+    /// .thread_name_fn(|| {
+    ///    static ATOMIC_ID: AtomicUsize = AtomicUsize::new(0);
+    ///    let id = ATOMIC_ID.fetch_add(1, Ordering::SeqCst);
+    ///    format!("my-pool-{}", id)
+    /// })
+    /// .build();
+    /// # }
```

Comment:
nit: can we indent this?
```suggestion
    /// let rt = runtime::Builder::new()
    ///     .thread_name_fn(|| {
    ///        static ATOMIC_ID: AtomicUsize = AtomicUsize::new(0);
    ///        let id = ATOMIC_ID.fetch_add(1, Ordering::SeqCst);
    ///        format!("my-pool-{}", id)
    ///     })
    ///     .build();
    /// # }
```